Don't suggest `42u8.into()`, suggest "change the literal to `42u16`

[Given the following code](https://play.rust-lang.org/?gist=e60e99a855bb39eeda0fd3c2eb1383e4&version=nightly&mode=debug&edition=2015)

```rust
fn foo(_: u16) {}
fn main() {
    foo(8u8);
}
```

we currently suggest using `.into()` to turn the `u8` literal into a `u16`

```rust
error[E0308]: mismatched types
 --> src/main.rs:3:9
  |
3 |     foo(8u8);
  |         ^^^ expected u16, found u8
help: you can cast an `u8` to `u16`, which will zero-extend the source value
  |
3 |     foo(8u8.into());
  |         ^^^^^^^^^^
```

We should suggest changing the literal directly when appropriate in the following way:

```rust
error[E0308]: mismatched types
 --> src/main.rs:3:9
  |
3 |     foo(8u8);
  |         ^^^ expected u16, found u8
help: change the type of the numeric literal from `u8` to `u16`
  |
3 |     foo(8u16);
  |         ^^^^
```

This has the added benefit that we can look at the literal and see if it would also fit when reducing the size (from 16 to 8, for example) if the literal would fit in the smaller type.

The current suggestion is implemented in [`check_for_cast`]( https://github.com/estebank/rust/blob/47a8eb7c4e24b61a8a9ab4eaff60ef65291aaa56/src/librustc_typeck/check/demand.rs#L315) and would require evaluating the `expr` to see if it is a literal with an explicit type suffix or not.

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -422,7 +422,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             let suffix_suggestion = format!(
                 "{}{}{}{}",
                 if needs_paren { "(" } else { "" },
-                src,
+                {
+                    // 42u8
+                    // ^^
+                    let lit_offset = src.len() - checked_ty.to_string().len();
+                    &src[..lit_offset]
+                },
                 expected_ty,
                 if needs_paren { ")" } else { "" },
             );

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -422,12 +422,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             let suffix_suggestion = format!(
                 "{}{}{}{}",
                 if needs_paren { "(" } else { "" },
-                {
-                    // 42u8
-                    // ^^
-                    let lit_offset = src.len() - checked_ty.to_string().len();
-                    &src[..lit_offset]
-                },
+                src.trim_right_matches(&checked_ty.to_string()),
                 expected_ty,
                 if needs_paren { ")" } else { "" },
             );

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -415,24 +415,54 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                           src,
                                           if needs_paren { ")" } else { "" },
                                           expected_ty);
-            let into_suggestion = format!("{}{}{}.into()",
-                                          if needs_paren { "(" } else { "" },
-                                          src,
-                                          if needs_paren { ")" } else { "" });
-            let suffix_suggestion = format!(
-                "{}{}{}{}",
+            let into_suggestion = format!(
+                "{}{}{}.into()",
                 if needs_paren { "(" } else { "" },
-                src.trim_right_matches(&checked_ty.to_string()),
-                expected_ty,
+                src,
                 if needs_paren { ")" } else { "" },
             );
-
-            let is_suffixed = |expr: &hir::Expr| {
+            let literal_is_ty_suffixed = |expr: &hir::Expr| {
                 if let hir::ExprKind::Lit(lit) = &expr.node {
                     lit.node.is_suffixed()
                 } else {
                     false
                 }
+            };
+
+            let into_sugg = into_suggestion.clone();
+            let suggest_to_change_suffix_or_into = |err: &mut DiagnosticBuilder,
+                                                    note: Option<&str>| {
+                let suggest_msg = if literal_is_ty_suffixed(expr) {
+                    format!(
+                        "change the type of the numeric literal from `{}` to `{}`",
+                        checked_ty,
+                        expected_ty,
+                    )
+                } else {
+                    match note {
+                        Some(note) => format!("{}, which {}", msg, note),
+                        _ => format!("{} in a lossless way", msg),
+                    }
+                };
+
+                let suffix_suggestion = format!(
+                    "{}{}{}{}",
+                    if needs_paren { "(" } else { "" },
+                    src.trim_right_matches(&checked_ty.to_string()),
+                    expected_ty,
+                    if needs_paren { ")" } else { "" },
+                );
+
+                err.span_suggestion_with_applicability(
+                    expr.span,
+                    &suggest_msg,
+                    if literal_is_ty_suffixed(expr) {
+                        suffix_suggestion
+                    } else {
+                        into_sugg
+                    },
+                    Applicability::MachineApplicable,
+                );
             };
 
             match (&expected_ty.sty, &checked_ty.sty) {
@@ -459,25 +489,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             }
                         }
                         _ => {
-                            if is_suffixed(expr) {
-                                err.span_suggestion_with_applicability(
-                                    expr.span,
-                                    &format!(
-                                        "change the type of the numeric literal from `{}` to `{}`",
-                                        checked_ty,
-                                        expected_ty,
-                                    ),
-                                    suffix_suggestion,
-                                    Applicability::MaybeIncorrect,
-                                );
-                            } else {
-                                err.span_suggestion_with_applicability(
-                                    expr.span,
-                                    &format!("{}, which {}", msg, will_sign_extend),
-                                    into_suggestion,
-                                    Applicability::MachineApplicable
-                                );
-                            }
+                            suggest_to_change_suffix_or_into(
+                                err,
+                                Some(will_sign_extend),
+                            );
                         }
                     }
                     true
@@ -505,25 +520,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             }
                         }
                         _ => {
-                            if is_suffixed(expr) {
-                                err.span_suggestion_with_applicability(
-                                    expr.span,
-                                    &format!(
-                                        "change the type of the numeric literal from `{}` to `{}`",
-                                        checked_ty,
-                                        expected_ty,
-                                    ),
-                                    suffix_suggestion,
-                                    Applicability::MaybeIncorrect,
-                                );
-                            } else {
-                                err.span_suggestion_with_applicability(
-                                    expr.span,
-                                    &format!("{}, which {}", msg, will_zero_extend),
-                                    into_suggestion,
-                                    Applicability::MachineApplicable
-                                );
-                            }
+                           suggest_to_change_suffix_or_into(
+                               err,
+                               Some(will_zero_extend),
+                           );
                         }
                     }
                     true
@@ -624,25 +624,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 }
                 (&ty::Float(ref exp), &ty::Float(ref found)) => {
                     if found.bit_width() < exp.bit_width() {
-                        if is_suffixed(expr) {
-                            err.span_suggestion_with_applicability(
-                                expr.span,
-                                &format!(
-                                    "change the type of the numeric literal from `{}` to `{}`",
-                                    checked_ty,
-                                    expected_ty,
-                                ),
-                                suffix_suggestion,
-                                Applicability::MaybeIncorrect,
-                            );
-                        } else {
-                            err.span_suggestion_with_applicability(
-                                expr.span,
-                                &format!("{} in a lossless way", msg),
-                                into_suggestion,
-                                Applicability::MachineApplicable
-                            );
-                        }
+                       suggest_to_change_suffix_or_into(
+                           err,
+                           None,
+                       );
                     } else if can_cast {
                         err.span_suggestion_with_applicability(
                             expr.span,

--- a/src/librustc_typeck/check/demand.rs
+++ b/src/librustc_typeck/check/demand.rs
@@ -419,6 +419,21 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                           if needs_paren { "(" } else { "" },
                                           src,
                                           if needs_paren { ")" } else { "" });
+            let suffix_suggestion = format!(
+                "{}{}{}{}",
+                if needs_paren { "(" } else { "" },
+                src,
+                expected_ty,
+                if needs_paren { ")" } else { "" },
+            );
+
+            let is_suffixed = |expr: &hir::Expr| {
+                if let hir::ExprKind::Lit(lit) = &expr.node {
+                    lit.node.is_suffixed()
+                } else {
+                    false
+                }
+            };
 
             match (&expected_ty.sty, &checked_ty.sty) {
                 (&ty::Int(ref exp), &ty::Int(ref found)) => {
@@ -444,12 +459,25 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             }
                         }
                         _ => {
-                            err.span_suggestion_with_applicability(
-                                expr.span,
-                                &format!("{}, which {}", msg, will_sign_extend),
-                                into_suggestion,
-                                Applicability::MachineApplicable
-                            );
+                            if is_suffixed(expr) {
+                                err.span_suggestion_with_applicability(
+                                    expr.span,
+                                    &format!(
+                                        "change the type of the numeric literal from `{}` to `{}`",
+                                        checked_ty,
+                                        expected_ty,
+                                    ),
+                                    suffix_suggestion,
+                                    Applicability::MaybeIncorrect,
+                                );
+                            } else {
+                                err.span_suggestion_with_applicability(
+                                    expr.span,
+                                    &format!("{}, which {}", msg, will_sign_extend),
+                                    into_suggestion,
+                                    Applicability::MachineApplicable
+                                );
+                            }
                         }
                     }
                     true
@@ -477,12 +505,25 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             }
                         }
                         _ => {
-                            err.span_suggestion_with_applicability(
-                                expr.span,
-                                &format!("{}, which {}", msg, will_zero_extend),
-                                into_suggestion,
-                                Applicability::MachineApplicable
-                            );
+                            if is_suffixed(expr) {
+                                err.span_suggestion_with_applicability(
+                                    expr.span,
+                                    &format!(
+                                        "change the type of the numeric literal from `{}` to `{}`",
+                                        checked_ty,
+                                        expected_ty,
+                                    ),
+                                    suffix_suggestion,
+                                    Applicability::MaybeIncorrect,
+                                );
+                            } else {
+                                err.span_suggestion_with_applicability(
+                                    expr.span,
+                                    &format!("{}, which {}", msg, will_zero_extend),
+                                    into_suggestion,
+                                    Applicability::MachineApplicable
+                                );
+                            }
                         }
                     }
                     true
@@ -583,12 +624,25 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 }
                 (&ty::Float(ref exp), &ty::Float(ref found)) => {
                     if found.bit_width() < exp.bit_width() {
-                        err.span_suggestion_with_applicability(
-                            expr.span,
-                            &format!("{} in a lossless way", msg),
-                            into_suggestion,
-                            Applicability::MachineApplicable
-                        );
+                        if is_suffixed(expr) {
+                            err.span_suggestion_with_applicability(
+                                expr.span,
+                                &format!(
+                                    "change the type of the numeric literal from `{}` to `{}`",
+                                    checked_ty,
+                                    expected_ty,
+                                ),
+                                suffix_suggestion,
+                                Applicability::MaybeIncorrect,
+                            );
+                        } else {
+                            err.span_suggestion_with_applicability(
+                                expr.span,
+                                &format!("{} in a lossless way", msg),
+                                into_suggestion,
+                                Applicability::MachineApplicable
+                            );
+                        }
                     } else if can_cast {
                         err.span_suggestion_with_applicability(
                             expr.span,

--- a/src/test/ui/mismatched_types/numeric-literal-cast.rs
+++ b/src/test/ui/mismatched_types/numeric-literal-cast.rs
@@ -1,0 +1,15 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn foo(_: u16) {}
+fn main() {
+    foo(8u8);
+}
+

--- a/src/test/ui/mismatched_types/numeric-literal-cast.rs
+++ b/src/test/ui/mismatched_types/numeric-literal-cast.rs
@@ -9,7 +9,12 @@
 // except according to those terms.
 
 fn foo(_: u16) {}
+fn foo1(_: f64) {}
+fn foo2(_: i32) {}
+
 fn main() {
-    foo(8u8);
+    foo(1u8);
+    foo1(2f32);
+    foo2(3i16);
 }
 

--- a/src/test/ui/mismatched_types/numeric-literal-cast.stderr
+++ b/src/test/ui/mismatched_types/numeric-literal-cast.stderr
@@ -1,0 +1,13 @@
+error[E0308]: mismatched types
+  --> $DIR/numeric-literal-cast.rs:13:9
+   |
+LL |     foo(8u8);
+   |         ^^^ expected u16, found u8
+help: change the type of the numeric literal from `u8` to `u16`
+   |
+LL |     foo(8u16);
+   |         ^^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

--- a/src/test/ui/mismatched_types/numeric-literal-cast.stderr
+++ b/src/test/ui/mismatched_types/numeric-literal-cast.stderr
@@ -1,13 +1,33 @@
 error[E0308]: mismatched types
-  --> $DIR/numeric-literal-cast.rs:13:9
+  --> $DIR/numeric-literal-cast.rs:16:9
    |
-LL |     foo(8u8);
+LL |     foo(1u8);
    |         ^^^ expected u16, found u8
 help: change the type of the numeric literal from `u8` to `u16`
    |
-LL |     foo(8u16);
+LL |     foo(1u16);
    |         ^^^^
 
-error: aborting due to previous error
+error[E0308]: mismatched types
+  --> $DIR/numeric-literal-cast.rs:17:10
+   |
+LL |     foo1(2f32);
+   |          ^^^^ expected f64, found f32
+help: change the type of the numeric literal from `f32` to `f64`
+   |
+LL |     foo1(2f64);
+   |          ^^^^
+
+error[E0308]: mismatched types
+  --> $DIR/numeric-literal-cast.rs:18:10
+   |
+LL |     foo2(3i16);
+   |          ^^^^ expected i32, found i16
+help: change the type of the numeric literal from `i16` to `i32`
+   |
+LL |     foo2(3i32);
+   |          ^^^^
+
+error: aborting due to 3 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

